src/Hosting/Server.IntegrationTesting/src/Deployers/SelfHostDeployer.cs

```diff
@@ -55,23 +56,33 @@ public override async Task<DeploymentResult> DeployAsync()
                     DotnetPublish();
                 }
 
-                var hintUrl = TestUriHelper.BuildTestUri(
-                    DeploymentParameters.ServerType,
-                    DeploymentParameters.Scheme,
-                    DeploymentParameters.ApplicationBaseUriHint,
-                    DeploymentParameters.StatusMessagesEnabled);
-
                 // Launch the host process.
-                var (actualUrl, hostExitToken) = await StartSelfHostAsync(hintUrl);
-
-                Logger.LogInformation("Application ready at URL: {appUrl}", actualUrl);
-
-                return new DeploymentResult(
-                    LoggerFactory,
-                    DeploymentParameters,
-                    applicationBaseUri: actualUrl.ToString(),
-                    contentRoot: DeploymentParameters.PublishApplicationBeforeDeployment ? DeploymentParameters.PublishedApplicationRootPath : DeploymentParameters.ApplicationPath,
-                    hostShutdownToken: hostExitToken);
+                for (var i = 0; i < RetryCount; i++)
+                {
+                    try
+                    {
+                        var hintUrl = TestUriHelper.BuildTestUri(
+                            DeploymentParameters.ServerType,
+                            DeploymentParameters.Scheme,
+                            DeploymentParameters.ApplicationBaseUriHint,
+                            DeploymentParameters.StatusMessagesEnabled);
+                        var (actualUrl, hostExitToken) = await StartSelfHostAsync(hintUrl);
+
+                        Logger.LogInformation("Application ready at URL: {appUrl}", actualUrl);
+
+                        return new DeploymentResult(
+                            LoggerFactory,
+                            DeploymentParameters,
+                            applicationBaseUri: actualUrl.ToString(),
+                            contentRoot: DeploymentParameters.PublishApplicationBeforeDeployment ? DeploymentParameters.PublishedApplicationRootPath : DeploymentParameters.ApplicationPath,
+                            hostShutdownToken: hostExitToken);
+                    }
+                    catch (Exception ex)
+                    {
+                        Logger.LogWarning($"Application failed to start, exception: {ex.Message}");
```

Comment:
I'm a little leary that the retry isn't specific to a known error (e.g. port conflicts). This could end up hiding real product issues. Logged warnings aren't ever looked at unless the test fails all retries.

Reply:
What would you recommend instead? I don't have any other alternatives that to retry for port conflicts?

Reply:
Retrying is fine, but we should only do it if we can detect the specific failure. I don't see how you can do that from this point in the code, you'd have to scrape the logs or console output? Or catch the exception in the test app and set a specific exit code?

The Kestrel tests use port 0 and scrape the console output, right?

The HttpSys in-proc functional tests start the server in a loop and catch that specific exception. Maybe we could fake port 0 for HttpSys by wrapping IServer in the test app?

Reply:
> The Kestrel tests use port 0 and scrape the console output, right?

Both scrape console output for the port.

> The HttpSys in-proc functional tests start the server in a loop and catch that specific exception. 

You are talking about https://github.com/aspnet/AspNetCore/blob/06d7fe73a995000ce79007e03785ee77d4dbb508/src/Servers/HttpSys/test/FunctionalTests/Utilities.cs#L109, correct?

> Maybe we could fake port 0 for HttpSys by wrapping IServer in the test app?

That would require all new applications that want to deploy HttpSys to wrap IServer. I'd rather have the fix in the deployer itself.

Reply:
> > The HttpSys in-proc functional tests start the server in a loop and catch that specific exception.
> 
> You are talking about
> 
> https://github.com/aspnet/AspNetCore/blob/06d7fe73a995000ce79007e03785ee77d4dbb508/src/Servers/HttpSys/test/FunctionalTests/Utilities.cs#L109
> 
> , correct?

Yes

> > Maybe we could fake port 0 for HttpSys by wrapping IServer in the test app?
> 
> That would require all new applications that want to deploy HttpSys to wrap IServer. 

There are only a handfull of tests that use the deployer with HttpSys, and that wrapper could be shared.

> I'd rather have the fix in the deployer itself.

The deployer doesn't have the granularity, unless you try scraping the log / console output for errors?

Reply:
Fixing the app rather than the deployer is a ticking time bomb for introducing flakiness again. Let's say I wanted to 

> The deployer doesn't have the granularity, unless you try scraping the log / console output for errors?

 I am suggesting scraping logs.

Reply:
See the latest commit. Thoughts?